### PR TITLE
Add TEST_FAIL macro for tests

### PR DESCRIPTION
**Pull request by AgathiyanB**
## Description

The TEST_ASSERT macro was being used in places where we wanted to just assert false, using the pattern `TEST_ASSERT(!"message")` which expanded to a tautology in the if statement, so the new macro ASSERT_FALSE is being used in these instances.

## PR checklist

Please tick as appropriate and edit the reasons (e.g.: "backport: not needed because this is a new feature")

- [x] **changelog** not required as no new functionality
- [x] **backport** #7940 
- [x] **tests** these are changes to the test implementation

### CODE REVIEW

**Comment by gilles-peskine-arm** [comment]
This needs to be backported. Generally, all changes to the test framework should be backported, unless they're in a support function that's only applicable to the development branch (e.g. a support function for bignum core). The reason is that we want to be able to backport new test cases quickly when we add a non-regression test for a bug fix.

File: tests/include/test/macros.h [hunk 1]
```diff
@@ -73,7 +83,7 @@
 #define TEST_EQUAL(expr1, expr2)                                      \
     do {                                                                \
         if (!mbedtls_test_equal( #expr1 " == " #expr2, __LINE__, __FILE__, \
-                                 expr1, expr2))                      \
+                                 (unsigned long long) (expr1), (unsigned long long) (expr2)))                      \
```
Comment (gilles-peskine-arm) [inline tests/include/test/macros.h:86]: Please don't combine unrelated changes in the same commit.

Why this change anyway?

**Comment by tom-cosgrove-arm** [comment]
Naming: would something like `TEST_FAIL()` be better, we're not really asserting anything, rather, we are stating that the tests are failing.

Also note that the names of these macros are inconsistent: I tried to tidy this up in #6253 but that stalled, but I would like to start progressing that again, so can we figure out the best names for these?

File: tests/include/test/macros.h (hunk 1 above)
Comment (AgathiyanB) [reply tests/include/test/macros.h]: It was needed for another PR as IAR was complaining about implicit type conversion and mbedtls_test_equal was defined to have arguments of type unsigned long long.

Comment (gilles-peskine-arm) [reply tests/include/test/macros.h]: > IAR was complaining about implicit type conversion and mbedtls_test_equal was defined to have arguments of type unsigned long long.

That's a valid argument, but it belongs in its own commit.

Though I don't like that it causes `TEST_EQUAL(p1, p2)` to compile when `p1` and `p2` are pointers. It's not catastrophic since on most platforms it will compare the pointers for equality, but it can have both false positives and false negatives on “exotic” architectures, especially architectures with rich pointers. Can we find a way to say that converting signed integer types are ok, but pointers are not ok?

Comment (AgathiyanB) [reply tests/include/test/macros.h]: I'll have a look into that in #7935

File: tests/include/test/macros.h [hunk 2]
```diff
@@ -61,6 +61,16 @@
...
+/** This macro asserts fails the test with given output message.
+ *
+ * \param   MESSAGE The message to be outputed on assertion
+ */
+#define TEST_FAIL(MESSAGE)                           \
+    do {                                                  \
+        mbedtls_test_fail(MESSAGE, __LINE__, __FILE__);   \
+        goto exit;                                          \
+    } while (0)                                              \
+
```
Comment (tom-cosgrove-arm) [inline tests/include/test/macros.h:73]: (minor) for new code, and editing blocks of existing code, can we line up the `\`s on the right? I normally aim for one or two spaces from the rightmost line ending (i.e. would be two spaces from the `;` on the `mbedtls_test_fail()` line)

Comment (tom-cosgrove-arm) [reply tests/include/test/macros.h:73]: Other than that this is looking good

Comment (tom-cosgrove-arm) [reply tests/include/test/macros.h:73]: oh - also remove the trailing `\` on the `} while (0)` line, please

Comment (tom-cosgrove-arm) [reply tests/include/test/macros.h:73]: (minor) the trailing `\` on line 68 doesn't line up, but only make that change if another change is needed later

**Review by tom-cosgrove-arm — DISMISSED** [review]
LGTM

**Comment by tom-cosgrove-arm** [comment]
@gilles-peskine-arm are you reviewing or commenting on this PR?

**Comment by gilles-peskine-arm** [comment]
Just commenting

**Review by tom-cosgrove-arm — APPROVED** [review]
LGTM

**Review by gilles-peskine-arm — APPROVED** [review]
LGTM

**Comment by tom-cosgrove-arm** [comment]
Looks like a network connectivity error during CI check in the merge queue

```
 > git fetch --no-tags --progress git@github.com:Mbed-TLS/mbedtls.git +refs/heads/gh-readonly-queue/development/pr-7936-33291ba35fc1e829826142357b36f85843a22083:refs/remotes/origin/gh-readonly-queue/development/pr-7936-33291ba35fc1e829826142357b36f85843a22083
ERROR: Error fetching remote repo 'origin'
hudson.plugins.git.GitException: Failed to fetch from git@github.com:Mbed-TLS/mbedtls.git
	at hudson.plugins.git.GitSCM.fetchFrom(GitSCM.java:888)
	at hudson.plugins.git.GitSCM.retrieveChanges(GitSCM.java:1155)
	at hudson.plugins.git.GitSCM.checkout(GitSCM.java:1186)
	at org.jenkinsci.plugins.workflow.steps.scm.SCMStep.checkout(SCMStep.java:125)
	at org.jenkinsci.plugins.workflow.steps.scm.SCMStep$StepExecutionImpl.run(SCMStep.java:93)
	at org.jenkinsci.plugins.workflow.steps.scm.SCMStep$StepExecutionImpl.run(SCMStep.java:80)
	at org.jenkinsci.plugins.workflow.steps.SynchronousNonBlockingStepExecution.lambda$start$0(SynchronousNonBlockingStepExecution.java:47)
	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at java.lang.Thread.run(Thread.java:748)
Caused by: hudson.plugins.git.GitException: Command "git fetch --no-tags --progress git@github.com:Mbed-TLS/mbedtls.git +refs/heads/gh-readonly-queue/development/pr-7936-33291ba35fc1e829826142357b36f85843a22083:refs/remotes/origin/gh-readonly-queue/development/pr-7936-33291ba35fc1e829826142357b36f85843a22083" returned status code 128:
stdout:
stderr: fatal: couldn't find remote ref refs/heads/gh-readonly-queue/development/pr-7936-33291ba35fc1e829826142357b36f85843a22083
```
